source/common/common/utility.cc

```diff
@@ -250,10 +250,40 @@ std::string StringUtil::escape(const std::string& source) {
 std::string AccessLogDateTimeFormatter::fromTime(const SystemTime& time) {
   static DateFormatter default_date_format("%Y-%m-%dT%H:%M:%S");
 
-  return fmt::format(
-      "{}.{:03d}Z", default_date_format.fromTime(time),
-      std::chrono::duration_cast<std::chrono::milliseconds>(time.time_since_epoch()).count() %
-          1000);
+  struct CachedTime {
+    std::chrono::seconds epoch_time_seconds;
+    std::string formatted_time;
+    bool initialized_{false};
+  };
+  static thread_local CachedTime cached_time;
+
+  std::chrono::milliseconds epoch_time_ms =
+      std::chrono::duration_cast<std::chrono::milliseconds>(time.time_since_epoch());
+
+  // The following string concatenation is equivalent to
+  //   std::string millis = fmt::format(".{:03d}Z", epoch_time_ms.count() % 1000)
+  // but the fmt::format call proved to be too expensive in benchmark testing.
+  std::string millis;
```

Comment:
Maybe create this with an explicit size since we know this in advance and can avoid any potential reallocation later.